REGRESSION (iOS 16): RELEASE_ASSERT in slow_path_wasm_loop_osr

krevis-figma opened the pull request:
#### a1935ab9387bb87df6e60b410ef8186080db0678
<pre>
REGRESSION (iOS 16): RELEASE_ASSERT in slow_path_wasm_loop_osr
<a href="https://bugs.webkit.org/show_bug.cgi?id=242294">https://bugs.webkit.org/show_bug.cgi?id=242294</a>

Reviewed by Yusuke Suzuki.

Fix a regression caused by <a href="https://bugs.webkit.org/show_bug.cgi?id=234587.">https://bugs.webkit.org/show_bug.cgi?id=234587.</a>

This affected only iOS, and only when the WebAssembly module was greater
than 10 MB. In that case, the option `webAssemblyBBQAirModeThreshold`
caused JavaScriptCore/wasm/WasmBBQPlan.cpp to fall back from Air to B3.
However, the change in 234587 in `loop_osr` didn&apos;t correctly handle that
fallback. Changed it to use the old code path in this case.

Test: None (to be added later by @justinmichaud)

* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::planGeneratesLoopOSREntrypoints): Extracted from compileFunction.
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL): If planGeneratesLoopOSREntrypoints, use the old code path.

Canonical link: <a href="https://commits.webkit.org/252167@main">https://commits.webkit.org/252167@main</a>
</pre>

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
```diff
@@ -60,6 +60,19 @@ BBQPlan::BBQPlan(Context* context, Ref<ModuleInformation> moduleInformation, uin
+    // as BBQ Air bloats such lengthy Wasm code and will consume a large amount of executable memory.
+    bool forceUsingB3 = false;
+    if (Options::webAssemblyBBQAirModeThreshold() && moduleInformation->codeSectionSize >= Options::webAssemblyBBQAirModeThreshold())
+        forceUsingB3 = true;
```
Let's do early return `return true;`

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
```diff
@@ -60,6 +60,19 @@ BBQPlan::BBQPlan(Context* context, Ref<ModuleInformation> moduleInformation, uin
+    // FIXME: Some webpages use very large Wasm module, and it exhausts all executable memory in ARM64 devices since the size of executable memory region is only limited to 128MB.
+    // The long term solution should be to introduce a Wasm interpreter. But as a short term solution, we introduce heuristics to switch back to BBQ B3 at the sacrifice of start-up time,
+    // as BBQ Air bloats such lengthy Wasm code and will consume a large amount of executable memory.
+    bool forceUsingB3 = false;
```
We do not need to have this variable.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
```diff
@@ -60,6 +60,19 @@ BBQPlan::BBQPlan(Context* context, Ref<ModuleInformation> moduleInformation, uin
+    if (Options::webAssemblyBBQAirModeThreshold() && moduleInformation->codeSectionSize >= Options::webAssemblyBBQAirModeThreshold())
+        forceUsingB3 = true;
+    else if (!Options::wasmBBQUsesAir())
+        forceUsingB3 = true;
```
We can make it
```
if (!Options::wasmBBQUsesAir())
    return true;
```

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
```diff
@@ -60,6 +60,19 @@ BBQPlan::BBQPlan(Context* context, Ref<ModuleInformation> moduleInformation, uin
+        forceUsingB3 = true;
+    else if (!Options::wasmBBQUsesAir())
+        forceUsingB3 = true;
+    return forceUsingB3;
```
Let's make it `return false` by converting the above if-else into early return.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQPlan.h:
```diff
@@ -75,6 +75,8 @@ class BBQPlan final : public EntryPlan {
         return Base::parseAndValidateModule(m_source.data(), m_source.size());
     }
 
+    static bool forceUsingB3(Ref<ModuleInformation>);
```
Let's take `const ModuleInformation&`

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
```diff
@@ -60,6 +60,19 @@ BBQPlan::BBQPlan(Context* context, Ref<ModuleInformation> moduleInformation, uin
     dataLogLnIf(WasmBBQPlanInternal::verbose, "Starting BBQ plan for ", functionIndex);
 }
 
+bool BBQPlan::forceUsingB3(Ref<ModuleInformation> moduleInformation)
```
Let's take `const ModuleInformation&`.

justinmichaud replied:
I would suggest naming this function planGeneratesLoopOSREntrypoints or something along those lines.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
```diff
@@ -214,16 +227,7 @@ std::unique_ptr<InternalFunction> BBQPlan::compileFunction(uint32_t functionInde
-        forceUsingB3 = true;
-
-    if (forceUsingB3)
+    if (forceUsingB3(m_moduleInformation))
```
Let's pass it by

```
if (forceUsingB3(m_moduleInformation.get()))
```

Constellation reviewed (approved):
r=me with comments.

krevis-figma committed:
REGRESSION (iOS 16): RELEASE_ASSERT in slow_path_wasm_loop_osr
https://bugs.webkit.org/show_bug.cgi?id=242294

Reviewed by Yusuke Suzuki.

Fix a regression caused by https://bugs.webkit.org/show_bug.cgi?id=234587.

This affected only iOS, and only when the WebAssembly module was greater
than 10 MB. In that case, the option `webAssemblyBBQAirModeThreshold`
caused JavaScriptCore/wasm/WasmBBQPlan.cpp to fall back from Air to B3.
However, the change in 234587 in `loop_osr` didn't correctly handle that
fallback. Changed it to use the old code path in this case.

Test: None (to be added later by @justinmichaud)

* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::planGeneratesLoopOSREntrypoints): Extracted from compileFunction.
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL): If planGeneratesLoopOSREntrypoints, use the old code path.

Canonical link: https://commits.webkit.org/252167@main
